packages/gatsby/src/schema/infer/inference-metadata.js

```diff
@@ -216,15 +220,20 @@ const _updateValueDescriptor = (
       })
       break
     }
-    case `listOfUnion`: {
+    case `relatedNode`:
+    case `relatedNodeList`: {
       const { nodes = {} } = typeInfo
```

Comment:
Wouldn't this process be faster with a `Map`? Seems like it's doing pseudo arbitrary property setting and getting. And presumably the number of nodes can grow at scale?

Reply:
You tell me %) Any idea how much faster it will be? I used objects because they are easier to debug, also keys are always strings here, and insertion order doesn't matter.

Reply:
Kind of depends on the size. If each `nodes` object is limited to, say, five keys. Then it all doesn't matter much. If the size scales with the size of the site (whether it be queries, pages, or images), then we should just run this against such a site and check. But in that case I predict a map will be faster.

The advantage of a map is that the engine will optimize hashing, expecting an arbitrary set of keys. For regular objects it will continuously try ot optimize, and fail to do so, which is expensive. Additionally, gathering all keys of a map is cheaper versus doing that on an arbitrary object.

There's a little overhead (not sure if that's still true today, actually) to using Maps vs objects so for very small objects I'm not sweating it too much.

Reply:
I suspect it will make a difference for govbook so lemme check that.

I think this step is separate from this PR so I don't see this PR blocked on it.